tests/integration/ha/test_ha.py

```diff
@@ -12,72 +12,323 @@
     )
     assert new_writes > writes, "writes not continuing to ZK"
 
     logger.info("Checking leader re-election...")
-    new_leader_name = await helpers.get_leader_name(ops_test)
+    new_leader_name = helpers.get_leader_name(ops_test, non_leader_hosts)
     assert new_leader_name != leader_name
 
+    logger.info("Continuing leader process...")
+    await helpers.send_control_signal(ops_test=ops_test, unit_name=leader_name, signal="SIGCONT")
+    await asyncio.sleep(30)  # letting writes continue while unit rejoins
+
     logger.info("Stopping continuous_writes...")
     cw.stop_continuous_writes()
+    await asyncio.sleep(10)  # buffer to ensure writes sync
 
     logger.info("Counting writes on surviving units...")
     last_write = cw.get_last_znode(
-        parent=parent, hosts=non_leader_hosts, username=username, password=password
+        parent=parent, hosts=non_leader_hosts, username=USERNAME, password=password
     )
     total_writes = cw.count_znodes(
-        parent=parent, hosts=non_leader_hosts, username=username, password=password
+        parent=parent, hosts=non_leader_hosts, username=USERNAME, password=password
     )
     assert last_write == total_writes
 
     logger.info("Checking old leader caught up...")
     last_write_leader = cw.get_last_znode(
-        parent=parent, hosts=leader_host, username=username, password=password
+        parent=parent, hosts=leader_host, username=USERNAME, password=password
     )
     total_writes_leader = cw.count_znodes(
-        parent=parent, hosts=leader_host, username=username, password=password
+        parent=parent, hosts=leader_host, username=USERNAME, password=password
     )
     assert last_write == last_write_leader
     assert total_writes == total_writes_leader
+
+
+@pytest.mark.abort_on_fail
+async def test_network_cut_self_heal(ops_test: OpsTest, request):
+    """Cuts and restores network on leader, cluster self-heals after IP change."""
+    hosts = helpers.get_hosts(ops_test)
+    leader_name = helpers.get_leader_name(ops_test, hosts)
+    leader_host = helpers.get_unit_host(ops_test, leader_name)
+    leader_machine_name = await helpers.get_unit_machine_name(ops_test, leader_name)
+    password = helpers.get_super_password(ops_test)
+    parent = request.node.name
+    non_leader_hosts = ",".join([host for host in hosts.split(",") if host != leader_host])
+
+    logger.info("Starting continuous_writes...")
+    cw.start_continuous_writes(parent=parent, hosts=hosts, username=USERNAME, password=password)
+    await asyncio.sleep(30)  # letting client set up and start writing
+
+    logger.info("Checking writes are running at all...")
+    assert cw.count_znodes(parent=parent, hosts=hosts, username=USERNAME, password=password)
+
+    logger.info("Cutting leader network...")
+    helpers.cut_unit_network(machine_name=leader_machine_name)
+    await asyncio.sleep(60)  # to give time for re-election, longer as network cut is weird
```

Comment:
**suggestion**: I saw on the mongo tests these sleep values being a literal with double the median reelection time or similar: [literal](https://github.com/canonical/mongodb-operator/blob/main/tests/integration/ha_tests/test_ha.py#L16-L17)
It looks more _sophisticated_ :)